### ISSUE TITLE
Does minWM inherit the multi-stage training from Causal-Forcing?

### ISSUE THREAD

**Issue by hywchina** [shengshu-ai/minWM#2]
I have learned a lot from your open-source project Causal-Forcing, and it is really an insightful and meaningful work.
May I ask: does the newly released minWM inherit and include the multi-stage training design from Causal-Forcing?
Thanks a lot for your great work!

**Comment by zhuhz22** [comment]
Yes exactly, minWM is the full-stack open-source training pipeline based on Causal Forcing. Our motivation for this project is that we found the community still lacks a complete full-stack pipeline covering data, training, and models, which hinders the development of this direction. As for the specific workflow, it is indeed based on Causal Forcing: we distill a bidirectional multi-step diffusion model into a real-time, interactive, few-step AR generator. More details will be released before May 17. Stay tuned.

**Comment by gracezhao1997** [comment]
Thanks for your interest! The initial version of the code has already been released, and we will continue to improve and update the repo in the following weeks. Stay tuned!

**Comment by INV-WZQ** [comment]
@zhuhz22 @gracezhao1997 Hello, great work! I am also working on world models, and I recently encountered an issue when adapting Causal Forcing++ for real-time generation.

After Stage 2 consistency distillation, I noticed that the generated videos tend to flicker, with noticeable background jitter. **My main question is whether such background jitter after Stage 2 is an expected phenomenon, or whether it usually indicates a potential implementation issue.**

I initially thought this might be further refined in Stage 3. However, in my experiments, after Stage 3, the generated videos become noticeably blurry, especially in the latter half of long video generation.

Since I integrated Causal Forcing++ into my own codebase, this issue may be caused by an implementation bug on my side. However, I have not been able to identify whether the problem comes from Stage 2 or Stage 3. Have you ever encountered a similar issue when developing or using Causal Forcing++? Any insights or suggestions on possible causes would be greatly appreciated.

**Comment by zhuhz22** [comment]
Would you like to share the exact video demo? Without actual videos, it seems hard to accurately decide whether it is a supposed phenomenon.

Generally speaking, after Stage 2 the videos are indeed supposed to be not very good, but it shouldn't be  too bad. Also, Stage 3 is supposed to solve the issue.

Also, would you like to share more training details, e.g., which model you are using, the specific batch size & learning rate, and so on.

**Comment by INV-WZQ** [comment]
@zhuhz22 Thank you for your reply. I would like to provide more details about my current setting and observations.

## Demos
I show 30-second demos here. The results become worse when generating videos longer than 30 seconds.

* Demo after Stage 2 (10k steps):
https://github.com/user-attachments/assets/3bf9ee49-96a5-4be3-819d-d35a57f39ed5

* Demo after Stage 3 (5k steps):
  https://github.com/user-attachments/assets/b7df02d4-c9ff-40d2-8c6a-98b84c475c19

## Setting

### Model

I am using a game world model, [ReactiveGWM](https://github.com/INV-WZQ/ReactiveGWM), which is based on Wan2.2-5B.

### Stage 2

```bash
stage: stage2

max_steps: 10000
lr: 2.0e-6
batch_size: 4

cd_discrete_N: 48
cd_guidance_scale: 3.0
cd_ema_decay: 0.99
cd_ema_start_step: 200
```

### Stage 3

```bash
stage: stage3

lr: 2.0e-6
lr_critic: 4.0e-7
batch_size: 4
dfake_gen_update_ratio: 5

dmd_score_frames: 26
dmd_long_rollout: true
dmd_rollout_min_frames: 26
dmd_rollout_max_frames: 101

real_guidance_scale: 3.0
fake_guidance_scale: 0.0
dmd_timestep_clamp: [20, 980]

dmd_train_kv_window: 16
dmd_train_sink: 2

dmd_ema_decay: 0.99
dmd_ema_start_step: 200
dmd_ema_shadow_device: cpu
```

**Comment by zhuhz22** [comment]
I notice that you train the model with `dmd_rollout_max_frames: 101`. So do you use LongLive-style marginal DMD (the student yields videos longer than the bidirectional score models in DMD)? Then for Stage 3, what's the bidirectional model used for the two score models?  I notice that `dmd_score_frames: 26`, dose it mean that the bidirectional model is trained on 26 frames?

If you use LongLive-style marginal DMD, I suggest strictly following their recipe: first train 26 frames (maybe 5s) Causal Forcing DMD, and as long as it yields good results, perform long video rollout.

BTW, I think your Stage 2 already yields good results, so maybe we should focus more on Stage 3.

**Comment by zhuhz22** [comment]
Wow, I just realized that you are the author of ReactiveGWM. Really great work! If more issues come up in later distillation, I would be very excited to have deeper discussions with you. If you don’t mind, you can also add me on WeChat at zhuhz22 for easier communication.
